src/linearization.jl

```diff
@@ -0,0 +1,197 @@
+"""
```

Comment:
General comment here: this file seems overly verbose. There's a lot of repeated code. Is there a way we can clean this up to make it easier to understand and maintain?